File: exercism/python/hexadecimal/hexadecimal.py

DIGITS = '0123456789abcdef'


def hexa(hx):
    hx = hx.lower()
    if any(d not in set(DIGITS) for d in hx):
        raise ValueError

    return sum(DIGITS.index(d) << (4 * i) for i, d in enumerate(reversed(hx)))

File: exercism/haskell/hexadecimal/Hexadecimal.hs
module Hexadecimal (hexToInt) where

import Data.Bits (shift)
import qualified Data.Map as M
import Data.Maybe (isNothing)

digits :: M.Map Char Int
digits = M.fromList $ zip "0123456789abcdef" [0..]

hexToInt :: String -> Int
hexToInt hex
  | any (isNothing . (`M.lookup` digits)) hex = 0
  | otherwise = sum $ map (uncurry shift) ranks
  where ranks = zip (map (digits M.!) $ reverse hex) [0,4..]
